### DISCUSSION TITLE
Testing could use some improvements

### DISCUSSION THREAD

**Discussion by ghostsquad** [go-task/task#699] [Ideas]
It looks like many of the tests for task are writing actual files to disk, and generally appear to be end-2-end tests and not unit tests. There's a lot of room to improve here, using dependency injection to inject in-memory filesystems, and other such things would allow more unit tests to be written, and also greatly improve code coverage.

**Comment by andreynering** [comment]
Hi @ghostsquad,

Using in-memory filesystem sounded like a good and simple idea to me initially, but after some thought, it may actually not be so simple after all.

Most files generated in tests are through commands like `echo` inside the Taskfiles, which means [mvdan/sh](https://github.com/mvdan/sh) is actually creating them in practice.

There's a chance that by using [`OpenHandler`](https://pkg.go.dev/mvdan.cc/sh/v3@v3.4.3/interp#OpenHandler) we would be able accomplish that, but I have the impression that this is not necessarily the intention of that API.

We could try, and if it shows to be too difficult/unreliable we would need to open an issue at [mvdan/sh](https://github.com/mvdan/sh) asking for an official support to in-memory filesystems. Even them it wouldn't be 100% because external commands would still write to the OS filesystem, of course. But for our use case (only calling builtins like `echo`) it should work in theory.

**Reply by ghostsquad** [reply]
Well, our tests don't need to write files, at least in theory.

**Reply by ghostsquad** [reply]
We can probably look at `gosh` for inspiration. At minimum, it allows passing stdin/out/err to the interpreter.

**Comment by pioio** [comment]
Did we actually test that in-memory filesystem would make things faster? If so, by how much?

In principle, any modern OS kernel should be transparently caching file writes/reads in memory by default anyway (unless client explicitly decides to wait for file to be written to the backing store before confirming the IO operation).

I might be wrong, but I would guess using in-memory filesystem for this wouldn't speed things up by much.
